[runtime] Make debugger log to ring buffer and maintain thread state model

alexanderkyte opened the pull request:
By more explicitly modeling the transitions made by the debugger as changes in a state machine, we will more easily be able to refactor the control flow and resource lifetimes in the debugger agent. This change should have little to no observable changes to someone using the debugger, but on crash they will have a dump that looks like 

https://gist.github.com/alexanderkyte/05243ad173691092e7cfaf6b7448eba7
or
https://gist.github.com/alexanderkyte/c20a74b496fd7d55c91c8d9c23e302da
.

This information comes from the ring buffer and the modeled state.

vargaz commented:
I'm not really a fan of this. Why do we have to move the debugger internal state to some other public header ? Can somebody really debug debugger problems using just this log file ? Why is it better than simply enabling logging with --debugger-agent=loglevel=10 ?

luhenry commented:
@vargaz I assume this is useful to people who do not want to enable the very verbose logging, and still help figuring out what could have gone wrong.

vargaz commented:
If we really want to do this, then we should add some functions to the debugger code which is called by this state code to collect the state that it needs, instead of just exporting everything.

luhenry commented:
The property of a state machine is that the state transitions can be done separately from the actions resulting of transitioning state. For example, if you take a look at `mono/utils/mono-threads-state-machine.c`, only the state transitions and logging are done in this file, not all the actions that are the result of transitioning state.

Moving all the code that transitions state to a different file doesn't make it a state machine, it's just moving code. So try to model the various states, the various transitions and the various actions resulting of these transitions; then you can come up with code to encode these. The state machine doesn't need to be thread safe nor atomic, that still can be done by the caller.

alexanderkyte commented:
I intended to change most of this moved code shortly but it's easy enough to do that then. I'll rebase the old code back into debugger-agent.

luhenry commented on mono/mini/debugger-agent.c:
```diff
@@ -113,144 +115,6 @@ typedef struct {
 	gboolean setpgid;
 } AgentConfig;
 
-typedef struct
```
Why do these struct and define need to move? The state-machine shouldn't have any knowledge of what the various actions are, or you have a leaky abstraction.

alexanderkyte replied:
When the structured logger takes a more enum-based view of the commands and events, a lot of these types will have to be public.

vargaz replied:
Can't the logging be added inside debugger-agent.c ? We are already doing text logging there.

alexanderkyte replied:
These specific named types would have to be made visible in any refactoring of the debugger making any real changes, because they have to be exported in order to export DebuggerTlsData

alexanderkyte replied:
Why do we want the debugger to be a single, impenetrable file with untracked global variables determining state rather than factoring it into easy-to-read subsystems with logged interop points?

kumpera commented:
I find it particularly weird that we're getting what's essentially 2 logging systems for the debugger.

It's fine to have a flight-recorder that only dumps on failure and that usually means it needs to be low overhead.

I think a similar thing could be accomplished by augmenting how sdb logging is done by:

1) making it an integral part of mono's logging/tracing infrastructure
2) improve mono's logging/tracing infrastructure to have finer controls and a flight recorder (which means we might need to support structured logging).

DavidKarlas commented:
@monojenkins build pkg

DavidKarlas commented:
@monojenkins build pkg

DavidKarlas commented:
@monojenkins build pkg

alexanderkyte commented:
@monojenkins build pkg

alexanderkyte commented:
@kumpera 

I think that it's no more strange than how SGEN has two logging systems (the logs, and the binary protocol) and how the threading system has the logs you get with a compile-time change as well as the state machine + logs. 

I agree with you that we should bring these many disparate logging components throughout the runtime together under a couple of abstractions (verbose disk logger, lightweight flight recorder, lightweight non-disk logger) with component-specific implementations. 

Whether you want this PR to be the one that creates the shared logging infrastructure to improve all of these many different logging systems throughout the runtime, or whether this should just add the thread-states-type flight recorder to the debugger, is a choice that I'll leave up to the reviewers. I, personally, think that this PR should stay simple. 

If I'm going to make this a full logging unification PR, we should have a meeting about which interfaces we want to use globally and what the characteristics about each should be.

alexanderkyte commented:
I moved the public types back into debugger-agent to make the diff look smaller and used some getters and setters to access the fields I needed. Most of the relevant types had already been broken out by Kumpera in the previous debugger-engine PR. 

Also moved the types into mini and made the code only take a separate lock. I do this because the non-debugger-agent-thread threads will never end up paused while holding the lock. Other threads often end up paused while holding the loader lock, making all of the normal metadata code that we use (as well as these helpers) unsafe to access from that debugger-agent thread. These dumpers will be called from that thread. 

A lot of stress testing on my machine shows the locking changes to be resilient to a race. I ran the debugger test suite 100x times without seeing any failures or crashes (aren't we supposed to have some races on master?). I got a deadlock/loader-lock-contention reproduction after 10 test runs previously. 

For anybody interested in reproducing locking problems related to taking the loader lock from debugger-agent.c, they happen incredibly often on Linux+x64 and not at all on OSX desktop.

alexanderkyte commented:
@vargaz I would appreciate a re-review of this code if you have the time

luhenry commented on mono/mini/debugger-agent.c:
```diff
@@ -858,6 +863,56 @@ debugger_agent_parse_options (char *options)
+{
+	DebuggerTlsData *tls = (DebuggerTlsData *) ref;
+
+	if (!tls)
```
Use `g_assertf(tls, "Cannot get state of null thread")`.

luhenry commented on mono/mini/debugger-agent.c:
```diff
@@ -10075,6 +10143,10 @@ debugger_thread (void *arg)
 
 	DEBUG_PRINTF (1, "[dbg] Agent thread started, pid=%p\n", (gpointer) (gsize) mono_native_thread_id_get ());
 
+	gchar *log_each_step_var = g_getenv ("MONO_DEBUGGER_LOG_AFTER_COMMAND");
```
Use `g_hasenv`.

luhenry commented on mono/mini/debugger-agent.c:
```diff
@@ -858,6 +863,56 @@ debugger_agent_parse_options (char *options)
+{
+	DebuggerTlsData *tls = (DebuggerTlsData *) ref;
+
+	if (!tls)
```
Use `g_assertf(tls, "Cannot get state of null thread")`.

luhenry commented on mono/mini/debugger-agent.c:
```diff
@@ -249,6 +251,9 @@ typedef struct {
+
+	// The state that the debugger expects the thread to be in
+	MonoDebuggerThreadState thread_state;
 } DebuggerTlsData;
```
To avoid having `mono_debugger_tls_thread_id (gpointer debuggerTlsData)` with `debuggerTlsData` being a `gpointer`, you can simply define `DebuggerTlsData` as an opaque struct in `debugger-agent.h` like `typedef struct _ DebuggerTlsData DebuggerTlsData;` and have `struct _DebuggerTlsData` at https://github.com/mono/mono/pull/8865/files#diff-cad67a3642d7f7e00161e3e81c29f25eL168. This will allow you to use `DebuggerTlsData*` anywhere you want, without having to expose its definition.

luhenry commented on mono/mini/debugger-engine.h:
```diff
@@ -154,6 +157,21 @@ typedef struct {
+mono_debugger_is_disconnected (void);
+
+intptr_t 
+mono_debugger_tls_thread_id (gpointer debuggerTlsData);
```
Coding convention: variable name is `debugger_tls_data` and return value is `gsize`.

luhenry commented on mono/mini/debugger-engine.h:
```diff
@@ -5,6 +5,9 @@
 #ifndef __MONO_DEBUGGER_ENGINE_H__
 #define __MONO_DEBUGGER_ENGINE_H__
 
+#include <mono/metadata/seq-points-data.h>
```
Why do we need to include `seq-points-data.h`? If that's necessary only for `debugger-agent.c` then we prefer to put it there to reduce the need to rebuild everything whenever we modify one header file.

alexanderkyte replied:
The debugger-agent.c failed to compile for me for a while, master was broken. I think a subsequent fix was pushed. This was my workaround. I'll remove it. Me pushing this was an accident.

luhenry commented on mono/mini/debugger-agent.c:
```diff
@@ -858,6 +863,56 @@ debugger_agent_parse_options (char *options)
+	return tls->thread_state;
+}
+
+intptr_t 
```
Use `gsize`.

alexanderkyte commented:
@monojenkins rebuild failed

alexanderkyte commented:
Looking at windows failure right now, a bit hard to see how it could be triggered by my changes. And it's not wrong, the code there is not typesafe. That's a type of assignment we often do though.

alexanderkyte commented:
I'm building and running this on my VS2017 install, having problems reproducing the CI failure

lambdageek commented on mono/mini/debugger-state-machine.h:
```diff
@@ -0,0 +1,71 @@
+#define __MONO_DEBUGGER_STATE_MACHINE__
+
+
+#include <config.h>
```
You shouldn't include config.h in a header.

luhenry replied:
I missed this one.

alexanderkyte replied:
Fixed, thanks

lambdageek commented on msvc/libmini-common.targets:
```diff
@@ -60,6 +60,8 @@
     <ClInclude Include="$(MonoSourceLocation)\mono\mini\debugger-engine.h" />
     <ClCompile Include="$(MonoSourceLocation)\mono\mini\debugger-engine.c" />
+    <ClCompile Include="$(MonoSourceLocation)\mono\mini\debugger-state-machine.c" />
+    <ClCompile Include="$(MonoSourceLocation)\mono\mini\debugger-state-machine.h" />
```
ClInclude, not ClCompile

alexanderkyte replied:
Oh my. Thank you so much. I think I looked at that file too many times and became blind to it.

jaykrell commented on mono/metadata/handle.h:
```diff
@@ -487,7 +487,7 @@ extern const MonoObjectHandle mono_null_value_handle;
 mono_handle_assign (MonoObjectHandleOut dest, MonoObjectHandle src)
 {
-	MONO_HANDLE_SUPPRESS (dest->__raw = (gpointer)(src ? MONO_HANDLE_RAW (src) : NULL));
+	MONO_HANDLE_SUPPRESS (dest->__raw = (MonoObject *)(src ? MONO_HANDLE_RAW (src) : NULL));
```
Change is good but why was it failing?

alexanderkyte replied:
Because on windows it's not valid to assign a gpointer rvalue to a MonoObject * lvalue, according to VS's output.

alexanderkyte commented:
@monojenkins build pkg
